### PR TITLE
IGNITE-20637 Implement recovery of distributed index building

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-20637

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java [hunk 1]
```diff
@@ -73,37 +73,38 @@
...
  *     <li>Then it is expected that the distributed index building event will be triggered for all partitions via
- *     {@link IndexBuildCompletionListener} (from {@link IndexBuilder#listen}); as a result of each of these events, the corresponding key
- *     {@code indexBuild.partition.<indexId>.<partitionId>} will be deleted from metastore.</li>
- *     <li>When all the {@code indexBuild.partition.<indexId>.<partitionId>} keys in the metastore are deleted,
- *     {@link MakeIndexAvailableCommand} will be executed for the corresponding index.</li>
- *     <li>At {@link CatalogEvent#INDEX_AVAILABLE}, key {@code indexBuild.inProgress.<indexId>} in the metastore will be deleted.</li>
+ *     {@link IndexBuildCompletionListener} (from {@link IndexBuilder#listen}); as a result of each of these events, the corresponding
+ *     {@link IndexManagementUtils#partitionBuildIndexMetastoreKey(int, int) partition building index key} will be deleted from
+ *     metastore.</li>
+ *     <li>When <b>all</b> the {@link IndexManagementUtils#partitionBuildIndexMetastoreKey(int, int) buildig index key partition} in the
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java:83]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java [hunk 2]
```diff
@@ -73,37 +73,38 @@
...
+ *     metastore are deleted, {@link MakeIndexAvailableCommand} will be executed for the corresponding index.</li>
+ *     <li>At {@link CatalogEvent#INDEX_AVAILABLE},
+ *     {@link IndexManagementUtils#inProgressBuildIndexMetastoreKey(int) in progress building index key} in the metastore will be
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java:86]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+     * @param revUpperBound Upper bound of metastore revision.
+     */
+    static boolean isMetastoreKeysPresentLocally(MetaStorageManager metastore, ByteArray keyPrefix, long revUpperBound) {
+        try (Cursor<Entry> cursor = metastore.prefixLocally(keyPrefix, revUpperBound)) {
+            return cursor.stream().map(Entry::value).anyMatch(Objects::nonNull);
+        }
+    }
+
+    /**
+     * Removes a {@code key} in the metastore if <b>present</b>.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:90]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+     *
+     * @param metaStorageManager Metastore manager.
+     * @param key Key to remove.
+     * @return Future result {@code true} if success remove was applied, otherwise {@code false}.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:94]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+     *
+     * @param key Key.
+     * @return Partition ID.
+     */
+    static int extractPartitionIdFromPartitionBuildIndexKey(String key) {
+        assert key.startsWith(PARTITION_BUILD_INDEX_KEY_PREFIX) : key;
+
+        String[] strings = key.split("\\.");
+
+        return Integer.parseInt(strings[3]);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:238]: I can write a test to immediately detect any problem, what do you think?

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+     *
+     * @param key Key.
+     * @return Index ID.
+     */
+    static int extractIndexIdFromPartitionBuildIndexKey(String key) {
+        assert key.startsWith(PARTITION_BUILD_INDEX_KEY_PREFIX) : key;
+
+        String[] strings = key.split("\\.");
+
+        return Integer.parseInt(strings[2]);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:252]: I can write a test to immediately detect any problem, what do you think?

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    /**
+     * Returns {@code true} if at least one key by prefix is <b>present</b> in the metastore locally.
+     *
+     * @param metastore Metastore manager.
+     * @param keyPrefix Key prefix to check.
+     * @param revUpperBound Upper bound of metastore revision.
+     */
+    static boolean isMetastoreKeysPresentLocally(MetaStorageManager metastore, ByteArray keyPrefix, long revUpperBound) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:83]: I prefer to leave it as it is.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    /**
+     * Returns {@code true} if the local node is no longer the primary replica at the timestamp of interest.
+     *
+     * @param primaryReplicaMeta Primary replica meta.
+     * @param localNode Local node.
+     * @param timestamp Timestamp of interest.
+     */
+    static boolean isLeaseExpire(ReplicaMeta primaryReplicaMeta, ClusterNode localNode, HybridTimestamp timestamp) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:262]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java [hunk 9]
```diff
@@ -167,6 +167,9 @@ public void stop() throws Exception {
      * <p>Example: when we start building an index, we will need {@link IndexStorage} (as well as storage {@link MvPartitionStorage}) to
      * build it and we can get them in {@link CatalogEvent#INDEX_CREATE} using this method.</p>
      *
+     * <p>During recovery, it is important to wait until the local node becomes a primary replica so that all index building commands are
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java:170]: For those who want to use this method or will be debugging in the part of the code where it is used.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+ * <p>Approximate recovery algorithm:</p>
+ * <ul>
+ *     <li>For registered indexes: <ul>
+ *         <li>If the new index did not have time to add
+ *         {@link IndexManagementUtils#putBuildIndexMetastoreKeysIfAbsent(MetaStorageManager, int, int) index building keys}, then add them
+ *         to the metastore if they are <b>absent</b>.</li>
+ *         <li>If there are no {@link IndexManagementUtils#partitionBuildIndexMetastoreKey(int, int) partition index building keys} left for
+ *         the index in the metastore, then we {@link MakeIndexAvailableCommand make the index available} in the catalog.</li>
+ *         <li>For partitions for which index building has not completed, we will wait until the primary replica is elected so that the
+ *         replication log will be applied. If after this we find out that the index has been build, we will remove the
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:68]: Everywhere I use “build” when it comes to indexes.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+ *         {@link IndexManagementUtils#partitionBuildIndexMetastoreKey(int, int) partition index building key} from the metastore if it is
+ *         <b>present</b>.</li>
+ *     </ul></li>
+ *     <li>For available indexes: <ul>
+ *         <li>Delete the {@link IndexManagementUtils#inProgressBuildIndexMetastoreKey(int) “index construction in progress” key} in the
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:73]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+    /**
+     * Recovers index availability.
+     *
+     * <p>NOTE: Should only be executed on node recovery.</p>
+     *
+     * @param recoveryRevision Metastore revision on recovery.
+     * @return Future of recovery execution.
+     */
+    public CompletableFuture<Void> recovery(long recoveryRevision) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:122]: Okay, I chose `recover`

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+            return putBuildIndexMetastoreKeysIfAbsent(
+                    metaStorageManager,
+                    indexId,
+                    getPartitionCountFromCatalog(catalogManager, indexId, catalogVersion)
+            );
+        }
+
+        if (!isMetastoreKeysPresentLocally(metaStorageManager, partitionBuildIndexMetastoreKeyPrefix(indexId), recoveryRevision)) {
+            // Without wait, since the deploy metastore watches will be only after the start of the components is completed and there will
+            // be dead lock.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:184]: fix it

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+        // TODO: IGNITE-20678 We need to compare by IDs: localNode.id().equals(primaryReplicaMeta.getLeaseholderId())
+        return !localNode.name().equals(primaryReplicaMeta.getLeaseholder()) && timestamp.after(primaryReplicaMeta.getExpirationTime());
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:264]: Let's ask @sanpwc

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+                .thenCompose(mvTableStorage -> inBusyLockAsync(busyLock, () -> {
+                    var replicationGroupId = new TablePartitionId(tableId, partitionId);
+
+                    return placementDriver.getPrimaryReplica(replicationGroupId, clock.now())
+                            .thenCompose(primaryReplicaMeta -> inBusyLockAsync(busyLock, () -> {
+                                ClusterNode localNode = clusterService.topologyService().localMember();
+
+                                if (primaryReplicaMeta == null || isLeaseExpire(primaryReplicaMeta, localNode, clock.now())) {
+                                    // Local node is not the primary replica, so we expect to elect the primary replica with applying the
+                                    // replication log. If a local node is elected, then IndexAvailabilityController will get rid of the
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:232]: > Is it about partition replication log (i.e. not MetaStorage replication log)? If yes, how does it relate to the MetaStorage keys? Are there some index build-related commands in the partition log?

About partition replication log

> If there are, then is it enough to wait for us to become a primary replica? Shouldn't we wait for the commands to be actually applied? (Here, partition safe time might help... or maybe not)

I consulted with guys who understand more about replicas, they said that it is enough to get confirmation that we are the primary replica, which means that the replication log of the partition has been applied.

> Also, what if there is no primary replica while a node starts, so it does nothing; but later it is elected as a primary replica. Do we listen for 'I've become a primary replice' event somewhere? Probably yes, but I cannot find it :)


You can see where the primary replica election event is processed in `org.apache.ignite.internal.index.IndexBuildController#onPrimaryReplicaElected`, which triggers the building of the index for partitions, i.e. adds a tasks to build the index in `org.apache.ignite.internal.index.IndexBuildController#scheduleBuildIndex`, when the index build is completed for the partition, we will receive an event about it and process it in `org.apache.ignite.internal.index.IndexAvailabilityController#onIndexBuildCompletionForPartition`.

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+        int indexId0 = indexId(catalogManager, INDEX_NAME + 0, clock);
+        int indexId1 = indexId(catalogManager, INDEX_NAME + 1, clock);
+
+        makeIndexAvailable(catalogManager, indexId0);
+        makeIndexAvailable(catalogManager, indexId1);
+
+        // Let's put the inProgressBuildIndexMetastoreKey for only one index in the metastore.
+        putInProgressBuildIndexMetastoreKeyInMetastore(indexId0);
+
+        // Restart the components and perform a recovery.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:154]: ok. remove its

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+        restartComponentsAndPerformRecovery();
+
+        // Let's do checks.
+        assertMetastoreKeyPresent(metaStorageManager, inProgressBuildIndexMetastoreKey(indexId));
+        assertTrue(indexDescriptor(catalogManager, INDEX_NAME, clock).available());
+    }
+
+    @Test
+    void testRemovePartitionBuildIndexMetastoreKeyForRegisteredIndex() throws Exception {
+        // Let's make preparations.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:183]: This messages does not bother you, do you want to delete it?

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java [hunk 18]
```diff
@@ -0,0 +1,399 @@
...
+        int partitionId = 0;
+
+        putInProgressBuildIndexMetastoreKeyInMetastore(indexId);
+        putPartitionBuildIndexMetastoreKeyInMetastore(indexId, partitionId);
+
+        TablePartitionId replicaGroupId = new TablePartitionId(tableId, partitionId);
+        ClusterNode localNode = new ClusterNodeImpl(NODE_NAME + "_ID", NODE_NAME, mock(NetworkAddress.class));
+        ReplicaMeta primaryReplicaMeta = newPrimaryReplicaMeta(localNode, replicaGroupId, clock.now(), clock.now().addPhysicalTime(1_000));
+
+        // An empty array on purpose.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:197]: How do you propose to deal with checks because they are also different.

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java [hunk 19]
```diff
@@ -0,0 +1,399 @@
...
+        int tableId = tableId(catalogManager, TABLE_NAME, clock);
+        int indexId = indexId(catalogManager, INDEX_NAME, clock);
+        int partitionId = 0;
+
+        putInProgressBuildIndexMetastoreKeyInMetastore(indexId);
+        putPartitionBuildIndexMetastoreKeyInMetastore(indexId, partitionId);
+
+        TablePartitionId replicaGroupId = new TablePartitionId(tableId, partitionId);
+        ClusterNode localNode = new ClusterNodeImpl(NODE_NAME + "_ID", NODE_NAME, mock(NetworkAddress.class));
+        ReplicaMeta primaryReplicaMeta = newPrimaryReplicaMeta(localNode, replicaGroupId, clock.now(), clock.now().addPhysicalTime(1_000));
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:195]: Added one day)

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java [hunk 20]
```diff
@@ -0,0 +1,399 @@
...
+
+        // An empty array on purpose.
+        setIndexStorageToIndexManager(replicaGroupId, indexId);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:198]: I don’t understand your proposal, they are simply preparing IndexStorage in IndexManager.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:83]: The problem here is that it's not clear whether we want to check that ALL keys are present, or just ANY of them. This looks important. Without this clarification, the reader has to look into the method code.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:238]: Both would be useful, but a fix in the code seems to be more useful. When the code is less fragile, it's always good.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+import org.apache.ignite.network.ClusterService;
+
+/**
+ * Component responsible for restoring the algorithm from {@link IndexAvailabilityController} if a node fails at some step.
+ *
+ * <p>Approximate recovery algorithm:</p>
+ * <ul>
+ *     <li>For registered indexes: <ul>
+ *         <li>If the new index did not have time to add
+ *         {@link IndexManagementUtils#putBuildIndexMetastoreKeysIfAbsent(MetaStorageManager, int, int) index building keys}, then add them
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:63]: In the first item it's enough to just name one key (the one about the index as a whole), others are always added with it.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:68]: 'has been build' is broken English. You must use 3rd form of the verb after 'has been'. 'build' is the 1st form, 'built' is the 3rd.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:232]: >that the replication log of the partition has been applied

Has been applied until which index? And how is this index related to the moment when we make the recovery?

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerTest.java [hunk 22]
```diff
@@ -140,7 +128,7 @@ void testMetastoreKeysAfterIndexCreate() throws Exception {
 
         int indexId = indexId(INDEX_NAME);
 
-        awaitTillGlobalMetastoreRevisionIsApplied();
+        awaitTillGlobalMetastoreRevisionIsApplied(metaStorageManager);
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerTest.java:131]: My suggestion leads to eliminating a non-needed detail in a few places. What advantages does your suggestion ('leave it as is') bring?

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:182]: Yes, let's remove it do avoid stating the obvious

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:199]: I was only walking about the preparatory step, the checks are fine.

**Review by rpuch — COMMENTED** [review]
.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:83]: I suggest: `isAnyMetastoreKeyPresentLocally`.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:238]: I don’t understand at the moment why my new code is more fragile than the previous one?
With tests, these fragility can be detected quickly.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:83]: We discussed it in person and chose the name I suggested.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:238]: We discussed in person that for now I’ll just add tests.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:252]: We discussed in person that for now I’ll just add tests.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:68]: Ok

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerTest.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerTest.java:131]: Okay, I added a method without arguments to this test class.

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:182]: Remove it

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:199]: I don’t understand what you are ultimately suggesting. =)

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:63]: I don’t agree that it’s better to specify only the first key; I don’t want to copy-paste an already existing javadoc.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:232]: During recovery, the partition uses its replication log, which includes commands for building the index.

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:198]: I suggest extracting lines 182-202 to a method that will have some parameters to tune its behavior so that it can be reused in the following tests

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java]: Well, `isLeaseExpired()` means that the node had the node was a primary replica AND the meta still says that it is a primary replica, but the lease has expired.

But this method actually checks for 'node is not a primary replica, OR we can't be sure that it is a primary replica'. This is something different. `isLeaseExpired()` seems misleading.

The reason I suggested to invert the logic was to get a simpler name like `stillPrimaryReplica()` or even `isPrimaryReplica()`. Both are short and more accurate than the current name.

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:199]: This is a part of the previous proposal, let's drop this one and concentrate on the previous.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java]: I choose `isPrimaryReplica()`.

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java:199]: ok

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java:232]: Could you please clarify that a primary replica, when we get it on recovery, is guaranteed to have the log of its group fully applied? Ideally, both here (close to the close that relies on this guarantee) and in the class javadoc (around line 68).

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorerTest.java]: Since you suggested I couldn’t do it, there would be a lot of parameters that I don’t like, I tried to combine some code areas.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityControllerRestorer.java]: I tried to modify comments.

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java [hunk 23]
```diff
@@ -259,8 +259,8 @@ static int extractIndexIdFromPartitionBuildIndexKey(String key) {
      * @param localNode Local node.
      * @param timestamp Timestamp of interest.
      */
-    static boolean isLeaseExpire(ReplicaMeta primaryReplicaMeta, ClusterNode localNode, HybridTimestamp timestamp) {
+    static boolean isPrimaryReplica(ReplicaMeta primaryReplicaMeta, ClusterNode localNode, HybridTimestamp timestamp) {
```
Comment (rpuch) [inline modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:262]: The method name is now inverted (isPrimaryReplica), but the code of the method checks the opposite. It seems that it was forgotten that the logic of the method (and the logic of the call) need to be inverted too

Comment (tkalkirill) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java:262]: Fix it

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexManagementUtilsTest.java [hunk 24]
```diff
@@ -66,4 +86,36 @@ void testExtractIndexIdFromPartitionBuildIndexKey() {
...
+        HybridTimestamp startTime = clock.now();
+        long dayInMillis = TimeUnit.DAYS.toMillis(1);
+
+        ReplicaMeta replicaMeta = newPrimaryReplicaMeta(LOCAL_NODE, replicaGroupId, startTime, startTime.addPhysicalTime(dayInMillis));
+
+        assertTrue(isPrimaryReplica(replicaMeta, LOCAL_NODE, clock.now()));
+    }
+
+    @Test
+    void testIsPrimaryReplicaFalse() {
```
Comment (rpuch) [inline modules/index/src/test/java/org/apache/ignite/internal/index/IndexManagementUtilsTest.java:103]: This test seems to actually have 2 test cases inside, I would split it into 2, but it's up to you

**Review by rpuch — APPROVED** [review]
Great job! Let's either wait for @sanpwc for his opinion on clock skew in `between()`, or postpone this (maybe file a ticket to address it later).

File: modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/index/src/main/java/org/apache/ignite/internal/index/IndexManagementUtils.java]: In given case both options will work properly and thus I'd rather use compareTo because it allows to proceed with index build process a little bit longer.

Generally, it depends on what you need.
In case of using after() within comparison you will have cluster wide decision of whether lease is expired or not. Meaning that if any node X is saying that lease is expired then any other node will also say that lease is expired, despite the fact that for some specific nodes it'll be possible to have local time less then lease expiration timestamp.

In case of  using simple compareTo it'll be local decision of whether lease is expired or not.

File: modules/index/src/test/java/org/apache/ignite/internal/index/IndexManagementUtilsTest.java (hunk 24 above)
Comment (tkalkirill) [reply modules/index/src/test/java/org/apache/ignite/internal/index/IndexManagementUtilsTest.java:103]: I'll leave it as it is.
